src/main/java/org/sagebionetworks/web/client/widget/entity/menu/v3/EntityActionMenuImpl.java

```diff
@@ -0,0 +1,178 @@
+package org.sagebionetworks.web.client.widget.entity.menu.v3;
+
+import com.google.gwt.user.client.ui.IsWidget;
+import com.google.gwt.user.client.ui.Widget;
+import com.google.inject.Inject;
+import java.util.List;
+import java.util.Map;
+import org.sagebionetworks.repo.model.EntityType;
+import org.sagebionetworks.web.client.jsinterop.ReactMouseEvent;
+import org.sagebionetworks.web.client.jsinterop.entity.actionmenu.ActionConfiguration;
+import org.sagebionetworks.web.client.jsinterop.entity.actionmenu.EntityActionMenuDropdownConfiguration;
+import org.sagebionetworks.web.client.jsinterop.entity.actionmenu.EntityActionMenuDropdownMap;
+import org.sagebionetworks.web.client.jsinterop.entity.actionmenu.EntityActionMenuLayout;
+import org.sagebionetworks.web.client.widget.entity.menu.v2.Action;
+
+public class EntityActionMenuImpl
+  implements EntityActionMenu, EntityActionMenuView.Presenter {
+
+  private EntityActionMenuLayout currentLayout;
+  private final EntityActionMenuDropdownMap dropdownMenuConfigurations;
+  private final Map<Action, ActionConfiguration> actionConfigurations;
+
+  private final EntityActionMenuView view;
+  private boolean isLoading;
+
+  @Inject
+  public EntityActionMenuImpl(EntityActionMenuView view) {
+    this.view = view;
+    this.view.setPresenter(this);
+    actionConfigurations =
+      DefaultActionConfigurationUtil.getDefaultActionConfiguration();
+    dropdownMenuConfigurations =
+      EntityActionMenuDropdownMap.create(
+        EntityActionMenuDropdownConfiguration.create(true, "", false),
+        EntityActionMenuDropdownConfiguration.create(true, "", false)
+      );
+    currentLayout =
+      DefaultEntityActionMenuLayoutUtil.getLayout(EntityType.project);
+
+    synchronizeView();
+  }
+
+  @Override
+  public void setActionListener(
+    Action action,
+    ActionListenerV2 actionListener
+  ) {
+    actionConfigurations.get(action).clearActionListeners();
+    actionConfigurations.get(action).setHref(null);
+    addActionListener(action, actionListener);
+  }
+
+  @Override
+  public void addActionListener(
+    Action action,
+    ActionListenerV2 actionListener
+  ) {
+    actionConfigurations.get(action).addActionListener(actionListener);
+  }
+
+  @Override
+  public void setActionHref(Action action, String href) {
+    actionConfigurations.get(action).clearActionListeners();
+    addActionHref(action, href);
+  }
+
+  @Override
+  public void addActionHref(Action action, String href) {
+    actionConfigurations.get(action).setHref(href);
+    synchronizeView();
```

Comment:
Pretty much every change calls `synchronizeView` which re-renders the React component.